### ISSUE TITLE
Lint/NestedMethodDefinition does not know about instance_exec and friends

### ISSUE THREAD

**Issue by buehmann** [bbatsov/rubocop#4185]
Lint/NestedMethodDefinition allows methods to be nested in cases like these:
```ruby
def works
  instance_eval do
    def bar; end
  end
  class_eval do
    def bar; end
  end
  module_eval do
    def bar; end
  end
end
```
On the other hand it complains about the following similar cases where the `_exec` variant of the `_eval` methods is used or a singleton class is opened via `class <<`:
```ruby
def missing
  instance_exec do
    def bar; end
  end
  class_exec do
    def bar; end
  end
  module_exec do
    def bar; end
  end
  class << self
    def bar; end
  end
end
```

```
Inspecting 1 file
W

Offenses:

lint_nested_method_definition.rb:1:1: C: Method has too many lines. [12/10]
def missing ...
^^^^^^^^^^^
lint_nested_method_definition.rb:3:5: W: Method definitions must not be nested. Use lambda instead.
    def bar; end
    ^^^^^^^^^^^^
lint_nested_method_definition.rb:6:5: W: Method definitions must not be nested. Use lambda instead.
    def bar; end
    ^^^^^^^^^^^^
lint_nested_method_definition.rb:9:5: W: Method definitions must not be nested. Use lambda instead.
    def bar; end
    ^^^^^^^^^^^^
lint_nested_method_definition.rb:12:5: W: Method definitions must not be nested. Use lambda instead.
    def bar; end
    ^^^^^^^^^^^^

1 file inspected, 5 offenses detected
```

## Expected behavior

I expect Lint/NestedMethodDefinition not to complain about these cases.

## RuboCop version

```
$ rubocop -V
0.48.0 (using Parser 2.4.0.0, running on ruby 2.3.1 x86_64-darwin15)
```

**Comment by buehmann** [comment]
@Drenmi @bbatsov Thanks a lot for the fix. One case is not covered though (I take the blame for not including it explicitly in the title of this issue but only in the description):
```ruby
class Foo
  def x(obj)
    class << obj
      def y
      end
    end
  end
end
```

```
bug.rb:4:7: W: Method definitions must not be nested. Use lambda instead.
      def y ...
      ^^^^^
```
Shall I open another issue for this?

**Comment by Drenmi** [comment]
@buehmann: This is a bit of a cross cutting concern. I'm working on a node extension for `def` nodes in which I attempt to move many of the scoping checks into a single place.

Please feel free to open an issue though. The more traceability, the better. 🙂

**Comment by buehmann** [comment]
Okay, thanks. Done.
